### PR TITLE
Support PIN of unlimited length

### PR DESCRIPTION
**Pull request by andrewkozlik**
This PR implements https://github.com/trezor/trezor-firmware/issues/1167 to provide support for an unlimited-length PIN and wipe code. For TT there is indeed no limit, but for T1 the PIN length needs to be limited due to the size of `msg_tiny` which needs to fit in a single USB packet. I picked 50 digits, because it's the same as the maximum passphrase length, it provides sufficient security (158 bits) and you can't fit much more into a single 64 byte USB packet. Although technically anything as large as 81 digits could make sense, it's not worth reworking the usb_tiny mechanism. Do we want to set a limit for the PIN length on the TT? 

Wallet and Suite (and presumably 3rd party wallets) will need to be updated to support longer PIN. I already created a [branch in mytrezor](https://github.com/satoshilabs/mytrezor/tree/pin-40-digits) earlier, but haven't tested it. @tsusanka please coordinate this with the other teams.

Note that since this involves a storage version upgrade, downgrading the device will cause it to get wiped. I bumped the minor version numbers and FIX_VERSIONs. There are ways to hack this together without needing a storage version upgrade but I would prefer to do this properly and generically, which is bound to be more future-proof.

In storage.c I introduced UNAUTH_VERSION_KEY, which is the unauthenticated plaintext variant of VERSION_KEY. It would be nicer to store this as an authenticated unencrypted entry which we don't support yet. I think when we last discussed this type of entry, the prevailing opinion was that the use case is so rare, that it should be solved by having a private and a public entry containing the same information, as above.

@onvej-sl please review the storage changes in d78a7da, 97f483a.
@matejcik please review the rest of the PR. (The vast majority is just technical changes, the interesting stuff might be the new upgrade tests and a UI improvement in core.)

Test cases for QA:
- Old wipe code still works immediately after firmware upgrade when unlocking for the first time (T1, TT).
- Old wipe code still works after firmware upgrade and a successful unlock (T1, TT).
- Test wipe code with and without SD-protect (TT only).
- Upgrade TT from 2.0.9 with PIN enabled.

### CODE REVIEW

**Review by onvej-sl — APPROVED** [review]

File: storage/tests/tests/test_random.py [hunk 1]
```diff
@@ -9,11 +9,13 @@
 class StorageComparison(RuleBasedStateMachine):
     def __init__(self):
         super(StorageComparison, self).__init__()
+        print("X")
```
[the remark this answers is not in the document]
Comment (andrewkozlik) [reply storage/tests/tests/test_random.py:12]: :-D definitely not. Thanks for noticing.

**Comment by matejcik** [comment]
there aren't any tests for non-numeric PINs

**Comment by matejcik** [comment]
> Do we want to set a limit for the PIN length on the TT?

I smell something like "kid breaks iPhone lock screen by typing for five minutes" if we don't, so I am in favor of setting a limit. 50 digits seems reasonable, same as passphrase and same as T1

**Comment by andrewkozlik** [comment]
> there aren't any tests for non-numeric PINs

Currently the user cannot enter a non-numeric PIN. On T1 it would be complicated to implement, but on TT it should be quite easy to replace the numeric-only keyboard with the existing passphrase keyboard. I remember asking about this at some point and IIRC the decision from the product team was that the PIN should be numeric-only. I guess there should be a non-numeric PIN test in the storage tests, but do you think there should also be such a unit test and/or device test for the TT?

> > Do we want to set a limit for the PIN length on the TT?
> 
> I smell something like "kid breaks iPhone lock screen by typing for five minutes" if we don't, so I am in favor of setting a limit. 50 digits seems reasonable, same as passphrase and same as T1

Agreed.

**Comment by matejcik** [comment]
> but do you think there should also be such a unit test and/or device test for the TT?

I would expect an unit test, yes. Device test is probably pointless, given that the user cannot actually enter a non-numeric value.

File: core/src/trezor/ui/components/tt/pin.py [hunk 2]
```diff
@@ -82,7 +106,7 @@ def __init__(
         prompt: str,
         subprompt: Optional[str],
         allow_cancel: bool = True,
-        maxlength: int = 9,
+        maxlength: Optional[int] = 50,
```
Comment (matejcik) [inline core/src/trezor/ui/components/tt/pin.py:109]: let's make this not optional; nobody should need actually unbounded PIN

**Comment by matejcik** [comment]
just one minor comment, otherwise basically LGTM

just to confirm:

* Are we intentionally internally-allowing alphanumeric PINs?
* Are we intentionally not internally limiting PIN length? This means that there's no device test for max PIN length on TT, as the only thing enforcing the length is the UI widget (which is sidestepped in device tests). This feels strange, but I can't imagine a scenario where it poses any sort of actual problem, so just confirming that this is really intentional.

**Comment by matejcik** [comment]
one more thing, the storage size is unaffected by PIN length, correct?

**Comment by andrewkozlik** [comment]
> Are we intentionally internally-allowing alphanumeric PINs?

Yes, I tried to make it as generic as possible, because I have come to regret not having done this in the first place. As I already commented above, there are ways to hack this feature together without needing a storage version upgrade but I would prefer to do this properly and generically, which is bound to be more future-proof.

> Are we intentionally not internally limiting PIN length? This means that there's no device test for max PIN length on TT, as the only thing enforcing the length is the UI widget (which is sidestepped in device tests). This feels strange, but I can't imagine a scenario where it poses any sort of actual problem, so just confirming that this is really intentional.

Yes. There seems to be no reason for storage to impose some limits. It doesn't need to allocate any space, just reads the data it is given and processes them via PBKDF2.

> the storage size is unaffected by PIN length, correct?

Correct. But it's affected by the wipe code length ...

**Review by matejcik — APPROVED** [review]

**Comment by andrewkozlik** [comment]
@matejcik's comment lead me to realize that maybe we should give some more thought to the length of the wipe code entry in the storage. The reason why the entry contains the actual wipe code is only for integrity checking of the entry itself. I see these options:
1. Explicitly impose a 50-digit limit on the wipe code length in the storage in addition to the 50-digit limit already imposed in core and legacy. Just to emphasize and remind us that the wipe code shouldn't be excessively long, whereas with the PIN the length has no effect on storage size.
2. Change the integrity check. For example, we could store a 16-byte digest `wipe_code_digest = SHA-256(wipe_code)[:16]` instead of the wipe code itself. So the entry would be `(wipe_code_digest || salt || HMAC-SHA256-64(salt, wipe_code_digest))`. That way the size would be constant.
3. Leave as is.

I think going with Option 1 is fine, but I'd like to give it some thought still and hear your opinions before merging @onvej-sl, @matejcik. Note that if we ever want to change the format like in Option 2, then that would mean another storage upgrade, so let's be careful about what we decide.

**Review by prusnak — APPROVED** [review]

**Comment by matejcik** [comment]
to me (2) seems better, as the length of the wipe code won't affect other storage features.

however, given what wipe code does, and that the maximum would be 50, the difference is probably not significant

**Comment by onvej-sl** [comment]
I'm OK with both (1) and (2).

**Comment by onvej-sl** [comment]
Please, update https://github.com/trezor/trezor-firmware/blob/andrewkozlik/long-pin/docs/storage/index.md before you merge it.

**Comment by andrewkozlik** [comment]
I implemented Option 1:
> Explicitly impose a 50-digit limit on the wipe code length in the storage in addition to the 50-digit limit already imposed in core and legacy. Just to emphasize and remind us that the wipe code shouldn't be excessively long, whereas with the PIN the length has no effect on storage size.

I think the upgrade process would get more complicated with Option 2. But I can give it some more thought if necessary.

> Please, update https://github.com/trezor/trezor-firmware/blob/andrewkozlik/long-pin/docs/storage/index.md before you merge it.
> 
> > Prefix the entered PIN with a "1" digit in base 10 and convert the integer to 4 bytes in little endian byte order.

Good point. Done https://github.com/trezor/trezor-firmware/pull/1478/commits/2db656baec64a74e75c2e048754660314be5b3db.

> I've just realised that this is no longer true either:
> 
> > A 32-bit PVC would be sufficient to verify the PIN value, since there would be less than a 1 in 4 chance that there exists a false PIN, which has the same PVC as the correct PIN. Nevertheless, we decided to go with a 64-bit PVC to achieve a larger security margin. The chance that there exists a false PIN, which has the same PVC as the correct PIN, _then drops below 1 in 10^10_. The existence of a false PIN does not appear to pose a security weakness, since the false PIN cannot be used to decrypt the protected entries.
> 
> Are we OK with that?

This is what I put in the documentation:
> The 64 bit PVC means that there is less than a 1 in 10<sup>19</sup> chance that a wrong PIN will happen to have the same PVC as the correct PIN. The existence of false PINs does not pose a security weakness since a false PIN cannot be used to decrypt the protected entries.

The same applies to the wipe code. The worst that can happen is that the device will get wiped. To me 10<sup>-19</sup>, i.e. 2<sup>-64</sup>, seems OK as a margin for error.

**Pull request merged**